### ISSUE TITLE
Stateless worker deactivated when other silo is shut down

### ISSUE THREAD

**Issue by fsimonazzi** [dotnet/orleans#1837]
I tweaked the azure hello world sample for a proof of concept test with stateless workers. What we found was that some of these were deactivated when a different silo was shut down (e.g. when scaling down silo worker roles). This was unexpected, as these should be local to a silo as far as we know. 

Here's the verbose logs for both local directory and catalog involving the grain's activation and deactivation

```
2016-06-13T20:56:50.851Z	OrleansAzureSilos_IN_0	"EventName=""DirectWrite"" Message=""[2016-06-13 20:56:50.847 GMT    15	VERBOSE 	100000	Catalog	100.118.30.51:22222]	CreateGrainInstance *grn/B42DD7CD/00000000+S100.118.30.51:22222:203547394@a201669d	"""
2016-06-13T20:56:50.907Z	OrleansAzureSilos_IN_0	"EventName=""DirectWrite"" Message=""[2016-06-13 20:56:50.908 GMT    18	VERBOSE 	100511	Catalog	100.118.30.51:22222]	About to call HelloEnvironmentGrains.MyWorkerGrain grain's OnActivateAsync() method [Activation: S100.118.30.51:22222:203547394*grn/B42DD7CD/00000000+S100.118.30.51:22222:203547394@a201669d #GrainType=HelloEnvironmentGrains.MyWorkerGrain Placement=StatelessWorkerPlacement State=Activating]	"""
2016-06-13T20:56:50.921Z	OrleansAzureSilos_IN_0	"EventName=""DirectWrite"" Message=""[2016-06-13 20:56:50.926 GMT    18	VERBOSE 	100512	Catalog	100.118.30.51:22222]	Returned from calling HelloEnvironmentGrains.MyWorkerGrain grain's OnActivateAsync() method [Activation: S100.118.30.51:22222:203547394*grn/B42DD7CD/00000000+S100.118.30.51:22222:203547394@a201669d #GrainType=HelloEnvironmentGrains.MyWorkerGrain Placement=StatelessWorkerPlacement State=Activating]	"""
2016-06-13T20:56:50.949Z	OrleansAzureSilos_IN_0	"EventName=""DirectWrite"" Message=""[2016-06-13 20:56:50.954 GMT    19	VERBOSE 	100000	Catalog	100.118.30.51:22222]	InitActivation is done: S100.118.30.51:22222:203547394*grn/B42DD7CD/00000000+S100.118.30.51:22222:203547394@a201669d	"""

2016-06-13T21:18:31.502Z	OrleansAzureSilos_IN_0	"EventName=""DirectWrite"" Message=""[2016-06-13 21:18:31.502 GMT    15	VERBOSE-2 	100000	Orleans.GrainDirectory.LocalGrainDirectory	100.118.30.51:22222]	Silo S100.118.30.51:22222:203547394 calculated directory partition owner silo S100.118.44.13:22222:203548064 for grain *grn/B42DD7CD/00000000+S100.118.30.51:22222:203547394: -1326467783 --> 1975604084	"""
2016-06-13T21:18:31.502Z	OrleansAzureSilos_IN_0	"EventName=""MessageEvent"" Message=""[2016-06-13 21:18:31.502 GMT    15	INFO   	100524	Catalog	100.118.30.51:22222]	Catalog is deactivating 1 activations due to a failure of silo S100.118.44.13:22222:203548064/x75C15374, since it is a primary directory partiton to these grain ids.	"" TraceSource=""WaWorkerHost.exe"""
2016-06-13T21:18:31.534Z	OrleansAzureSilos_IN_0	"EventName=""DirectWrite"" Message=""[2016-06-13 21:18:31.533 GMT    15	VERBOSE 	100000	Catalog	100.118.30.51:22222]	DeactivateActivations: 1 activations.	"""
2016-06-13T21:18:31.553Z	OrleansAzureSilos_IN_0	"EventName=""MessageEvent"" Message=""[2016-06-13 21:18:31.549 GMT    15	INFO   	100541	Catalog	100.118.30.51:22222]	DeactivateActivations: total 1 to shutdown, out of them 1 promptly, 0 later when become idle and 0 are already being destroyed or invalid.	"" TraceSource=""WaWorkerHost.exe"""
2016-06-13T21:18:31.955Z	OrleansAzureSilos_IN_0	"EventName=""MessageEvent"" Message=""[2016-06-13 21:18:31.955 GMT    15	INFO   	100503	Catalog	100.118.30.51:22222]	Starting DestroyActivations #0 of 1 activations	"" TraceSource=""WaWorkerHost.exe"""
2016-06-13T21:18:32.047Z	OrleansAzureSilos_IN_0	"EventName=""DirectWrite"" Message=""[2016-06-13 21:18:32.051 GMT    18	VERBOSE 	100514	Catalog	100.118.30.51:22222]	About to call HelloEnvironmentGrains.MyWorkerGrain grain's OnDeactivateAsync() method [Activation: S100.118.30.51:22222:203547394*grn/B42DD7CD/00000000+S100.118.30.51:22222:203547394@a201669d #GrainType=HelloEnvironmentGrains.MyWorkerGrain Placement=StatelessWorkerPlacement State=Deactivating]	"""
2016-06-13T21:18:32.054Z	OrleansAzureSilos_IN_0	"EventName=""DirectWrite"" Message=""[2016-06-13 21:18:32.051 GMT    18	VERBOSE 	100515	Catalog	100.118.30.51:22222]	Returned from calling HelloEnvironmentGrains.MyWorkerGrain grain's OnDeactivateAsync() method [Activation: S100.118.30.51:22222:203547394*grn/B42DD7CD/00000000+S100.118.30.51:22222:203547394@a201669d #GrainType=HelloEnvironmentGrains.MyWorkerGrain Placement=StatelessWorkerPlacement State=Deactivating]	"""
2016-06-13T21:18:32.086Z	OrleansAzureSilos_IN_0	"EventName=""DirectWrite"" Message=""[2016-06-13 21:18:32.080 GMT    15	VERBOSE 	100000	Orleans.GrainDirectory.LocalGrainDirectory	100.118.30.51:22222]	Silo S100.118.30.51:22222:203547394 removed silo S100.118.44.13:22222:203548064	"""
```

Is this outcome expected?

**Comment by gabikliot** [comment]
Looks like a bug at a first glance. I will take a deeper look later on.

**Comment by fsimonazzi** [comment]
Thanks. I can provide the repro project if that would help - it's really just the azure hello world with a bootstrapping provider and a stateless worker grain.

**Comment by gabikliot** [comment]
No need for the repro. The original log entry was enough.
The fix is in #1838.

Good catch! Thank you.

**Comment by sergeybykov** [comment]
Fixed via https://github.com/dotnet/orleans/pull/1838.
